### PR TITLE
Vim mode toggle button

### PR DESCRIPTION
**Pull request by ron-popov**
Adds a vim mode toggle button.

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -32,7 +32,9 @@
   <script src="js/mode/ruby.js"></script>
   <script src="js/mode/rust.js"></script>
   <script src="js/mode/shell.js"></script>
-
+  
+  <script src="js/cookies.min.js"></script>
```
Comment (hermanzdosilovic) [inline index.html]: Please use [localStorage](https://www.w3schools.com/html/html5_webstorage.asp) instead of cookies.

File: index.html [hunk 2]
```diff
@@ -113,6 +115,9 @@
         <button class="btn btn-default" id="downloadSourceBtn" data-toggle="tooltip" data-placement="bottom" title="Download source">
           <span class="glyphicon glyphicon-download-alt" aria-hidden="true"></span>
         </button>
+        <button class="btn btn-primary" id="vimBtn">
```
Comment (hermanzdosilovic) [inline index.html]: [Checkbox](https://getbootstrap.com/docs/3.3/components/#input-groups-checkboxes-radios) would be more appropriate for this use-case.

File: js/ide.js [hunk 3]
```diff
@@ -54,6 +54,22 @@ function handleResult(data) {
   $runBtn.button("reset");
 };
 
+function toggleVim() {
+    var isVimModeActive = (Cookies.get("vimMode")) == "on";
+
+    // alert(Cookies.get("vimMode"));
```
Comment (hermanzdosilovic) [inline js/ide.js]: This is ok when you are testing your feature, but you should remove these before submitting PR.

File: js/ide.js [hunk 4]
```diff
@@ -54,6 +54,22 @@ function handleResult(data) {
...
+
+    if(isVimModeActive) {
+        // we need to disable vim mode.
```
Comment (hermanzdosilovic) [inline js/ide.js]: Line below is self documented, no need for explanation.

File: js/ide.js [hunk 5]
```diff
@@ -54,6 +54,22 @@ function handleResult(data) {
...
+        Cookies.set("vimMode" , "off");
+    } else {
+        // we need to enable vim mode
```
Comment (hermanzdosilovic) [inline js/ide.js]: Same as above.

File: js/ide.js [hunk 6]
```diff
@@ -204,7 +221,12 @@ $(document).ready(function() {
         cm.replaceSelection(spaces);
       }
     }
-  });
+  };
+
+  if(Cookies.get("vimMode") == "on")
```
Comment (hermanzdosilovic) [inline js/ide.js]: Please use parentheses (`{`, `}`) and space before `if` and `(`.

File: js/ide.js [hunk 7]
```diff
@@ -54,6 +54,22 @@ function handleResult(data) {
...
+
+    if(isVimModeActive) {
+        // we need to disable vim mode.
+        Cookies.set("vimMode" , "off");
+    } else {
+        // we need to enable vim mode
+        Cookies.set("vimMode" , "on");
+    }
+
+    alert("Please refresh for changes to take effect");
```
Comment (hermanzdosilovic) [inline js/ide.js]: You don't need to refresh, we don't want that.

You can use:
```
sourceEditor.options.keyMap = "vim";
// or
sourceEditor.options.keyMap = "default";
```

and changes would be visible right away.

**Comment by ron-popov** [comment]
I fixed the issues you commented on and it works but the design of the checkbox is not that great, should i push anyway ?

**Comment by hermanzdosilovic** [comment]
Try using [bootstraptoggle](http://www.bootstraptoggle.com/) to customize checkboxes. Push when you are ready :smile:

**Comment by ron-popov** [comment]
Toggling vim mode is a bit glitchy when changing from normal mode to insert mode.

**Comment by hermanzdosilovic** [comment]
I have merged your changes in another PR where I polished some things and added some addons.

**Pull request closed without merging**